bd-logger/src/log_replay.rs

```diff
@@ -283,15 +304,86 @@ impl ProcessingPipeline {
     self.workflows_engine.maybe_persist(block).await;
 
     if block {
-      Self::finish_blocking_log_processing(flush_buffers_tx, flush_stats_trigger, matching_buffers)
-        .await?;
+      Self::finish_blocking_log_processing(
+        &self.flush_buffers_tx,
+        flush_stats_trigger,
+        matching_buffers,
+      )
+      .await?;
     }
 
     Ok(log_replay_result)
   }
 
+  fn handle_common_pre_buffer_write(
+    capture_screenshot_handler: &CaptureScreenshotHandler,
+    triggered_flush_buffers_action_ids: &BTreeSet<Cow<'_, FlushBufferId>>,
+    capture_screenshot: bool,
+  ) {
+    if !triggered_flush_buffers_action_ids.is_empty() {
+      log::debug!("triggered flush buffer action IDs: {triggered_flush_buffers_action_ids:?}");
+    }
+
+    if capture_screenshot {
+      capture_screenshot_handler.capture_screenshot();
+    }
+  }
+
+  pub(crate) async fn process_state_change(
+    &mut self,
+    state_change: bd_state::StateChange,
+    state: &bd_state::Store,
+    now: OffsetDateTime,
+    session_id: &str,
+  ) -> LogReplayResult {
+    let state_reader = state.read().await;
+
+    let empty_set = TinySet::default();
+    let mut result = self.workflows_engine.process_event(
+      bd_workflows::workflow::WorkflowEvent::StateChange(&state_change),
+      &empty_set,
+      &state_reader,
+      now,
+    );
+
+    let log_replay_result = LogReplayResult {
+      logs_to_inject: std::mem::take(&mut result.logs_to_inject)
+        .into_values()
+        .collect(),
+      workflow_debug_state: result.workflow_debug_state,
+      engine_has_debug_workflows: result.has_debug_workflows,
+    };
+
+    log::debug!("processed {state_change:?} state change");
+
+    Self::handle_common_pre_buffer_write(
+      &self.capture_screenshot_handler,
+      &result.triggered_flush_buffers_action_ids,
+      result.capture_screenshot,
+    );
+
+    // In order to work with session capture we need there to be a log that can be emitted with the
+    // action ID for the flush action. Since state changes typically don't have associated logs, we
+    // need to rely on the mechanism to synthesize a log here.
```

Comment:
Total aside but I learned about this when I was doing the changes for the proto encoding. It would be nice to somehow handle this specifically vs. fake the log but out of scope.

Reply:
IMO ideally we shouldn't have to rely on logs at all for communicating captures but yeah out of scope